node/core/approval-voting/src/import.rs

```diff
@@ -0,0 +1,1749 @@
+///   * update the rolling session window
+///   * compute our assignments
+///   * import the block and candidates to the approval DB
+///   * and return information about all candidates imported under each block.
+///
+/// It is the responsibility of the caller to schedule wakeups for each block.
+pub(crate) async fn handle_new_head(
+	ctx: &mut impl SubsystemContext,
+	state: &mut State<impl DBReader>,
+	db_writer: &impl AuxStore,
+	head: Hash,
+	finalized_number: &Option<BlockNumber>,
+) -> SubsystemResult<Vec<BlockImportedCandidates>> {
+	// Update session info based on most recent head.
+
+	let header = {
+		let (h_tx, h_rx) = oneshot::channel();
+		ctx.send_message(ChainApiMessage::BlockHeader(head, h_tx).into()).await;
+
+		match h_rx.await? {
+			Err(e) => {
+				return Err(SubsystemError::with_origin("approval-voting", e));
+			}
+			Ok(None) => {
+				tracing::warn!(target: LOG_TARGET, "Missing header for new head {}", head);
+				return Ok(Vec::new());
+			}
+			Ok(Some(h)) => h
+		}
+	};
+
+	if let Err(SessionsUnavailable)
+		= cache_session_info_for_head(
+			ctx,
+			&mut state.session_window,
+			head,
+			&header,
+		).await?
+	{
+		tracing::warn!(
+			target: LOG_TARGET,
+			"Could not cache session info when processing head {:?}",
+			head,
+		);
+
+		return Ok(Vec::new())
+	}
+
+	// If we've just started the node and haven't yet received any finality notifications,
+	// we don't do any look-back. Approval voting is only for nodes were already online.
+	let finalized_number = finalized_number.unwrap_or(header.number.saturating_sub(1));
+
+	let new_blocks = determine_new_blocks(ctx, &state.db, head, &header, finalized_number)
+		.map_err(|e| SubsystemError::with_origin("approval-voting", e))
+		.await?;
+
+	let mut approval_meta: Vec<BlockApprovalMeta> = Vec::with_capacity(new_blocks.len());
+	let mut imported_candidates = Vec::with_capacity(new_blocks.len());
+
+	// `determine_new_blocks` gives us a vec in backwards order. we want to move forwards.
+	for (block_hash, block_header) in new_blocks.into_iter().rev() {
+		let env = ImportedBlockInfoEnv {
+			session_window: &state.session_window,
+			assignment_criteria: &*state.assignment_criteria,
+			keystore: &state.keystore,
+		};
+
+		let ImportedBlockInfo {
+			included_candidates,
+			session_index,
+			assignments,
+			n_validators,
+			relay_vrf_story,
+			slot,
+		} = match imported_block_info(ctx, env, block_hash, &block_header).await? {
+			Some(i) => i,
+			None => continue,
+		};
+
+		let candidate_entries = approval_db::v1::add_block_entry(
+			db_writer,
+			block_header.parent_hash,
+			block_header.number,
+			approval_db::v1::BlockEntry {
+				block_hash: block_hash,
+				session: session_index,
+				slot,
+				relay_vrf_story: relay_vrf_story.0,
+				candidates: included_candidates.iter()
+					.map(|(hash, _, core, _)| (*core, *hash)).collect(),
+				approved_bitfield: bitvec::bitvec![BitOrderLsb0, u8; 0; n_validators],
```

Comment:
why is it `n_validators` in length? shouldn't it be `included_candidates.len()`?